File: it-agents/skills/code-architect/references/code-examples/swift/Observer.swift
// Observer — subscription mechanism to notify subscribers of events.

protocol Observer: AnyObject {
    func update(value: Int)
}

final class Subject {
    private var observers: [Observer] = []
    private var value = 0 {
        didSet { notify() }
    }

    func subscribe(_ observer: Observer) {
        observers.append(observer)
    }

    func setValue(_ newValue: Int) {
        value = newValue
    }

    private func notify() {
        observers.forEach { $0.update(value: value) }
    }
}

final class LoggingObserver: Observer {
    let name: String
    init(name: String) { self.name = name }
    func update(value: Int) {
        print("\(name) saw new value: \(value)")
    }
}

func runDemo() {
    let subject = Subject()
    subject.subscribe(LoggingObserver(name: "A"))
    subject.subscribe(LoggingObserver(name: "B"))

    subject.setValue(10)
    subject.setValue(42)
}

runDemo()

File: it-agents/skills/code-architect/references/code-examples/python/observer.py
"""Observer — subscription mechanism to notify subscribers of events."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Observer(ABC):
    @abstractmethod
    def update(self, value: int) -> None:
        ...


class Subject:
    """Maintains subscribers and broadcasts state changes."""

    def __init__(self) -> None:
        self._observers: list[Observer] = []
        self._value = 0

    def subscribe(self, observer: Observer) -> None:
        self._observers.append(observer)

    def unsubscribe(self, observer: Observer) -> None:
        self._observers.remove(observer)

    def set_value(self, value: int) -> None:
        self._value = value
        for observer in self._observers:
            observer.update(value)


class LoggingObserver(Observer):
    def __init__(self, name: str) -> None:
        self._name = name

    def update(self, value: int) -> None:
        print(f"{self._name} saw value {value}")


if __name__ == "__main__":
    subject = Subject()
    subject.subscribe(LoggingObserver("A"))
    subject.subscribe(LoggingObserver("B"))
    subject.set_value(42)
